File: mobile/app/src/main/java/pi/binqr/binqr/ScanActivity.java
package pi.binqr.binqr;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Environment;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.*;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.ResultPoint;
import com.journeyapps.barcodescanner.*;

import java.io.File;
import java.io.FileOutputStream;
import java.io.UnsupportedEncodingException;
import java.util.*;

/**
 * This sample performs continuous scanning, displaying the barcode and source image whenever
 * a barcode is scanned.
 */
public class ScanActivity extends AppCompatActivity {
    private DecoratedBarcodeView barcodeView;
    private Map<Integer, QRCode> scannedParts;
    private List<CheckBox> checkBoxes;
    private Context context = this;

    private BarcodeCallback callback = new BarcodeCallback() {
        @Override
        public void barcodeResult(BarcodeResult result) {
            if(result.getText() == null) {
                return;
            }

            byte[] rawBytes = new byte[0];
            try {
                rawBytes = result.getText().getBytes("ISO-8859-1");
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }

            QRCode qrCode = QRCode.fromBytes(rawBytes);

            if (scannedParts.containsKey(qrCode.getMetadata().getNumber())) {
                return;
            }

            if (checkBoxes.size() == 0) {
                LinearLayout progress = (LinearLayout) findViewById(R.id.progress);
                for (int i = 0; i < qrCode.getMetadata().getQuantity(); i++) {
                    CheckBox checkBox = new CheckBox(context);
                    checkBox.setText(String.format(Locale.getDefault(), "%d", i + 1));
                    checkBox.setClickable(false);
                    progress.addView(checkBox);
                    checkBoxes.add(checkBox);
                }
                TextView first_scan_text = (TextView) findViewById(R.id.first_scan_text);
                first_scan_text.setVisibility(View.GONE);
            }

            checkBoxes.get(qrCode.getMetadata().getNumber() - 1).setChecked(true);

            scannedParts.put(qrCode.getMetadata().getNumber(), qrCode);
//            if ()
//            saveFile(rawBytes);
        }


        @Override
        public void possibleResultPoints(List<ResultPoint> resultPoints) {
        }
    };

    protected void saveFile(byte[]... bytes) {
        File file = new File(Environment.getExternalStorageDirectory(), "photo.jpg");

        if (file.exists()) {
            file.delete();
        }

        try {
            FileOutputStream fos=new FileOutputStream(file.getPath());

            fos.write(bytes[0]);
            fos.close();
        }
        catch (java.io.IOException e) {
            Log.e("PictureDemo", "Exception in photoCallback", e);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_scan);

        checkAndAskForPermissions();

        barcodeView = (DecoratedBarcodeView) findViewById(R.id.barcode_scanner);
        DecoderFactory decoderFactory = new DefaultDecoderFactory(EnumSet.of(BarcodeFormat.QR_CODE), null, "ISO-8859-1", false);
        barcodeView.getBarcodeView().setDecoderFactory(decoderFactory);
        barcodeView.decodeContinuous(callback);

        scannedParts = new HashMap<>();
        checkBoxes = new ArrayList<>();
    }

    private void checkAndAskForPermissions() {
        String[] neededPermissions = new String[]{
                Manifest.permission.CAMERA,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
        };
        Set<String> ungrantedPermissions = new HashSet<>();

        for (String neededPermission : neededPermissions) {
            if (ContextCompat.checkSelfPermission(this, neededPermission) != PackageManager.PERMISSION_GRANTED) {
                ungrantedPermissions.add(neededPermission);
            }
        }

        if (ungrantedPermissions.size() > 0) {
            ActivityCompat.requestPermissions(this, ungrantedPermissions.toArray(new String[]{}), 0);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        barcodeView.resume();
    }

    @Override
    protected void onPause() {
        super.onPause();

        barcodeView.pause();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        return barcodeView.onKeyDown(keyCode, event) || super.onKeyDown(keyCode, event);
    }
}

File: mobile/app/src/main/java/pi/binqr/binqr/ScanActivity.kt
package pi.binqr.binqr

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.View
import android.widget.*
import com.google.zxing.BarcodeFormat
import com.google.zxing.ResultPoint
import com.journeyapps.barcodescanner.*

import java.io.UnsupportedEncodingException
import java.util.*

/**
 * This sample performs continuous scanning, displaying the barcode and source image whenever
 * a barcode is scanned.
 */
class ScanActivity : AppCompatActivity() {
    private var barcodeView: DecoratedBarcodeView? = null
    private var scannedParts: MutableSet<String>? = null

    private val callback = object : BarcodeCallback {
        override fun barcodeResult(result: BarcodeResult) {
            if (result.text == null || scannedParts!!.contains(result.text)) {
                // Prevent duplicate scans
                return
            }

            var rawBytes = ByteArray(0)
            try {
                rawBytes = result.text.toByteArray(charset("ISO-8859-1"))
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
            }

            scannedParts!!.add(result.text)
            if (true) {
                save(rawBytes)
            }
        }


        override fun possibleResultPoints(resultPoints: List<ResultPoint>) {}
    }

    private fun save(vararg bytes: ByteArray) {
        val intent = Intent(this, SaveActivity::class.java)
        intent.putExtra("file", bytes)
        startActivity(intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_scan)

        checkAndAskForPermissions()

        barcodeView = findViewById(R.id.barcode_scanner) as DecoratedBarcodeView?
        val decoderFactory = DefaultDecoderFactory(EnumSet.of(BarcodeFormat.QR_CODE), null, "ISO-8859-1", false)
        barcodeView!!.barcodeView.decoderFactory = decoderFactory
        barcodeView!!.decodeContinuous(callback)

        scannedParts = HashSet<String>()

        val progress = findViewById(R.id.progress) as LinearLayout?

        for (i in 1..5) {
            val checkBox = CheckBox(this)
            checkBox.text = i.toString()
            checkBox.isClickable = false
            progress!!.addView(checkBox)
        }

        val first_scan_text = findViewById(R.id.first_scan_text) as TextView?
        first_scan_text!!.visibility = View.GONE
    }

    private fun checkAndAskForPermissions() {
        val neededPermissions = arrayOf(Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        val ungrantedPermissions = neededPermissions.filterTo(HashSet<String>()) { ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED }

        if (ungrantedPermissions.size > 0) {
            ActivityCompat.requestPermissions(this, ungrantedPermissions.toArray(emptyArray()), 0)
        }
    }

    override fun onResume() {
        super.onResume()

        barcodeView!!.resume()
    }

    override fun onPause() {
        super.onPause()

        barcodeView!!.pause()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return barcodeView!!.onKeyDown(keyCode, event) || super.onKeyDown(keyCode, event)
    }
}
